src/mlpack/methods/cf/ncf.hpp

```diff
@@ -0,0 +1,238 @@
+/**
+ * @file ncf.hpp
+ * @author Haritha Nair
+ *
+ * Implementation of Neural Collaborative Filtering.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_NCF_NCF_HPP
+#define MLPACK_METHODS_NCF_NCF_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core.hpp>
+#include <mlpack/methods/ann/layer/layer.hpp>
+#include <mlpack/methods/ann/ffn.hpp>
+#include <ensmallen.hpp>
+
+namespace mlpack {
+namespace cf {
+/**
+ * This class implements Neural Collaborative Filtering (NCF). 
+ *
+ * The data matrix is a (user, item, rating) table.  Each column in the matrix
+ * should have three rows.  The first represents the user; the second represents
+ * the item; and the third represents the rating.  The user and item, while they
+ * are in a matrix that holds doubles, should hold integer (or size_t) values.
+ * The user and item indices are assumed to start at 0.
+ */
+class NCF
+{
+ public:
+  // Default constructor for NCF.
+  NCF()
+  {
+    // Nothing to do here.
+  }
+  /**
+   * Initialize the NCF object using any algorithm and optimizer, according to
+   * which a model will be trained. There are parameters that can
+   * be set; default values are provided for each of them.
+   *
+   * The provided dataset can shall be a coordinate list; that is, a 3-row
+   * matrix where each column corresponds to a (user, item, rating) entry in the
+   * matrix.
+   * @tparam OptimizerType The optimizer to train the network on.
+   * @param dataset Data matrix: dense matrix (coordinate lists).
+   * @param algorithm Algorithm to be used.
+   * @param optimizer Optimizer to be used to train the model.
+   * @param embedSize Size of embedding for each user and item being considered.
+   * @param neg Number of negative instances to consider per positive instance.
+   * @param epochs Number of epochs to train the model on.
+   * @param implicit Whether to convert data to implicit feedback rating form.
+   */
+  template<typename OptimizerType = ens::StandardSGD>
+  NCF(arma::mat& dataset,
+      std::string algorithm,
```

Comment:
do you know what is the type of algorithms and possible options here? we need to mention this in this doc.